# 1. Các thư viện cần thiết:

In [1]:
import pandas as pd
import numpy as np

# 2. Đọc dữ liệu lưu vào Dataframe:

In [2]:
# Đọc tệp CSV mà không có header
netflix_df = pd.read_csv('../data/netflix_titles.csv', encoding='ISO-8859-1', skiprows= 0)
netflix_df.set_index('show_id', inplace=True) # Chọn cột show_id làm index
netflix_df.head()

type                  title         director  \
show_id                                                    
s1         Movie   Dick Johnson Is Dead  Kirsten Johnson   
s2       TV Show          Blood & Water              NaN   
s3       TV Show              Ganglands  Julien Leclercq   
s4       TV Show  Jailbirds New Orleans              NaN   
s5       TV Show           Kota Factory              NaN   

                                                      cast        country  \
show_id                                                                     
s1                                                     NaN  United States   
s2       Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
s3       Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
s4                                                     NaN            NaN   
s5       Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

                 date_added  release_year rating   duration  \
show_id                                                       
s1       September 25, 2021          2020  PG-13     90 min   
s2       September 24, 2021          2021  TV-MA  2 Seasons   
s3       September 24, 2021          2021  TV-MA   1 Season   
s4       September 24, 2021          2021  TV-MA   1 Season   
s5       September 24, 2021          2021  TV-MA  2 Seasons   

                                                 listed_in  ... Unnamed: 16  \
show_id                                                     ...               
s1                                           Documentaries  ...         NaN   
s2         International TV Shows, TV Dramas, TV Mysteries  ...         NaN   
s3       Crime TV Shows, International TV Shows, TV Act...  ...         NaN   
s4                                  Docuseries, Reality TV  ...         NaN   
s5       International TV Shows, Romantic TV Shows, TV ...  ...         NaN   

         Unnamed: 17  Unnamed: 18  Unnamed: 19  Unnamed: 20  Unnamed: 21  \
show_id                                                                    
s1               NaN          NaN          NaN          NaN          NaN   
s2               NaN          NaN          NaN          NaN          NaN   
s3               NaN          NaN          NaN          NaN          NaN   
s4               NaN          NaN          NaN          NaN          NaN   
s5               NaN          NaN          NaN          NaN          NaN   

         Unnamed: 22  Unnamed: 23  Unnamed: 24  Unnamed: 25  
show_id                                                      
s1               NaN          NaN          NaN          NaN  
s2               NaN          NaN          NaN          NaN  
s3               NaN          NaN          NaN          NaN  
s4               NaN          NaN          NaN          NaN  
s5               NaN          NaN          NaN          NaN  

[5 rows x 25 columns]

- Giải thích cho việc sử dụng cột `show_id` trong tập dữ liệu gốc làm index cho các dòng dữ liệu: Do các giá trị trong cột `show_id` là các giá trị độc lập, không trùng lắp nên sẽ thuận tiện cho việc truy xuất từng bộ phim/chương trình truyền hình thông qua ID của chúng.

# 3. Khám phá dữ liêu:

### 1. Số dòng, số cột của tập dữ liệu:

In [3]:
# Số dòng, số cột
row, column = netflix_df.shape
print(f'Số dòng của tập dữ liệu: {row}')
print(f'Số cột của tập dữ liệu: {column}')

Số dòng của tập dữ liệu: 8809
Số cột của tập dữ liệu: 25


- Sau khi quan sát, nhóm nhận thấy những cột cuối là những cột trống nên ở đây nhóm sẽ thực hiện xoá các cột trống.

In [4]:
# Xoá đi những cột có tất cả dòng có giá trị NaN
netflix_df = netflix_df.dropna(how='all', axis=1)
netflix_df.head()

type                  title         director  \
show_id                                                    
s1         Movie   Dick Johnson Is Dead  Kirsten Johnson   
s2       TV Show          Blood & Water              NaN   
s3       TV Show              Ganglands  Julien Leclercq   
s4       TV Show  Jailbirds New Orleans              NaN   
s5       TV Show           Kota Factory              NaN   

                                                      cast        country  \
show_id                                                                     
s1                                                     NaN  United States   
s2       Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
s3       Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
s4                                                     NaN            NaN   
s5       Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

                 date_added  release_year rating   duration  \
show_id                                                       
s1       September 25, 2021          2020  PG-13     90 min   
s2       September 24, 2021          2021  TV-MA  2 Seasons   
s3       September 24, 2021          2021  TV-MA   1 Season   
s4       September 24, 2021          2021  TV-MA   1 Season   
s5       September 24, 2021          2021  TV-MA  2 Seasons   

                                                 listed_in  \
show_id                                                      
s1                                           Documentaries   
s2         International TV Shows, TV Dramas, TV Mysteries   
s3       Crime TV Shows, International TV Shows, TV Act...   
s4                                  Docuseries, Reality TV   
s5       International TV Shows, Romantic TV Shows, TV ...   

                                               description  
show_id                                                     
s1       As her father nears the end of his life, filmm...  
s2       After crossing paths at a party, a Cape Town t...  
s3       To protect his family from a powerful drug lor...  
s4       Feuds, flirtations and toilet talk go down amo...  
s5       In a city of coaching centers known to train I...

In [5]:
# Số dòng, số cột sau khi loại bỏ các cột trống
row, column = netflix_df.shape
print(f'Số dòng của tập dữ liệu sau khi loại bỏ các cột trống: {row}')
print(f'Số cột của tập dữ liệu sau khi loại bỏ các cột trống: {column}')

Số dòng của tập dữ liệu sau khi loại bỏ các cột trống: 8809
Số cột của tập dữ liệu sau khi loại bỏ các cột trống: 11


### 2. Ý nghĩa mỗi dòng/cột dữ liệu:

#### Ý nghĩa mỗi dòng dữ liệu:

- Mỗi dòng dữ liệu cung cấp thông tin chi tiết về một bộ phim hoặc chương trình truyền hình, với các nội dung riêng biệt và rõ ràng.

#### Ý nghĩa mỗi cột dữ liệu:

- Mỗi cột dữ liệu cho biết những thông tin của một bộ phim hoặc chương trình truyền hình, bao gồm:
    - type: Danh mục có thể là bộ phim hoặc chương trình truyền hình.
    - title: Tựa đề của các bộ phim hoặc chương trình truyền hình.
    - director: Đạo diễn của các bộ phim/chương trình truyền hình. Một số mục sẽ có giá trị NULL, đặc biệt với các chương trình truyền hình.
    - cast: Danh sách các diễn viên chính.
    - country: Quốc gia là nơi sản xuất phim/chương trình truyền hình.
    - date_added: Ngày được đăng tải trên Netflix.
    - release_year: Năm sản xuất của bộ phim hoặc chương trình truyền hình.
    - Rating.
    - duration: Thời lượng của các bộ phim/chương trình truyền hình. Được đo bằng phút đối với các bộ phim và phần đối với các chương trình truyền hình.
    - listed_in: Thể loại mà tiêu đề thuộc về.
    - description: Mô tả ngắn về bộ phim/chương trình truyền hình.

### 3. Kiểm tra các dòng trùng lặp:

In [6]:
# Kiểm tra có tồn tại các dòng trùng lặp
row_duplicates = netflix_df[netflix_df.duplicated()]
print(f'Số dòng bị trùng lặp: {len(row_duplicates)}')

Số dòng bị trùng lặp: 3


- Như vậy trong tập dữ liệu có chứa các dòng trùng lặp.
- Tiến hành xoá các dòng trùng lặp.

In [7]:
# Thực hiện xoá các dòng trùng lặp
netflix_df = netflix_df.drop_duplicates()

In [8]:
# Kiểm tra còn tồn tại các dòng trùng lặp hay không
row_duplicates = netflix_df[netflix_df.duplicated()]
print(f'Số dòng bị trùng lặp sau khi thực hiện xoá: {len(row_duplicates)}')

Số dòng bị trùng lặp sau khi thực hiện xoá: 0


- Không còn tồn tại dòng nào bị trùng dữ liệu.

### 4. Kiểu dữ liệu của các cột:

In [9]:
# Kiểm tra kiểu dữ liệu của từng cột
netflix_df.dtypes

type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

- Có thể thấy giá trị của các cột gần như đã đúng hết. Nhưng để tiện sử dụng cho sau này thì nhóm sẽ thực hiện chuyển kiểu dữ liệu của cột dữ liệu `date_added` thành `datetime64` và cột `duration` vừa là đo bằng phút đối với các bộ phim vừa là phần đối với các chương trình truyền hình nên cần xử lý tách thành 2 cột riêng biệt để đảm bảo từng cột có một đơn vị đo khác nhau.

In [10]:
# Chuyển cột date_added thành kiểu datetime 64
netflix_df['date_added'] = netflix_df['date_added'].str.strip() # Chuyển thành chuỗi để loại bỏ khoảng trắng ở đầu và cuối của chuỗi thời gian
netflix_df['date_added'] = pd.to_datetime(netflix_df['date_added'])

In [11]:
# Tách cột duration thành 2 cột riêng biệt
movie_df = netflix_df[netflix_df['type'] == 'Movie']
movie_df = movie_df['duration'].str.extract(r'(\d+)\s*(\w+)')[0].rename('duration1')
tv_df = netflix_df[netflix_df['type'] == 'TV Show']
tv_df = tv_df['duration'].str.extract(r'(\d+)\s*(\w+)')[0].rename('season')
netflix_df = netflix_df.join([tv_df,movie_df])
netflix_df

type                  title         director  \
show_id                                                    
s1         Movie   Dick Johnson Is Dead  Kirsten Johnson   
s2       TV Show          Blood & Water              NaN   
s3       TV Show              Ganglands  Julien Leclercq   
s4       TV Show  Jailbirds New Orleans              NaN   
s5       TV Show           Kota Factory              NaN   
...          ...                    ...              ...   
s8805      Movie             Zombieland  Ruben Fleischer   
s8806      Movie                   Zoom     Peter Hewitt   
s8807      Movie                 Zubaan      Mozez Singh   
s8808    TV Show     Parasyte: The Grey     Yeon Sang-ho   
s8809      Movie                 Serena     Susanne Bier   

                                                      cast        country  \
show_id                                                                     
s1                                                     NaN  United States   
s2       Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
s3       Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
s4                                                     NaN            NaN   
s5       Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                    ...            ...   
s8805    Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
s8806    Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
s8807    Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   
s8808           Shin Hyun-been, Jeon Yeo-bin, Goo Kyo-hwan    South Korea   
s8809    Bradley Cooper, Jennifer Lawrence, Toby Jones,...  United States   

        date_added  release_year rating   duration  \
show_id                                              
s1      2021-09-25          2020  PG-13     90 min   
s2      2021-09-24          2021  TV-MA  2 Seasons   
s3      2021-09-24          2021  TV-MA   1 Season   
s4      2021-09-24          2021  TV-MA   1 Season   
s5      2021-09-24          2021  TV-MA  2 Seasons   
...            ...           ...    ...        ...   
s8805   2019-11-01          2009      R     88 min   
s8806   2020-01-11          2006     PG     88 min   
s8807   2019-03-02          2015  TV-14    111 min   
s8808   2024-04-05          2024  TV-MA   1 Season   
s8809   2024-04-05          2014      A    110 min   

                                                 listed_in  \
show_id                                                      
s1                                           Documentaries   
s2         International TV Shows, TV Dramas, TV Mysteries   
s3       Crime TV Shows, International TV Shows, TV Act...   
s4                                  Docuseries, Reality TV   
s5       International TV Shows, Romantic TV Shows, TV ...   
...                                                    ...   
s8805                              Comedies, Horror Movies   
s8806                   Children & Family Movies, Comedies   
s8807       Dramas, International Movies, Music & Musicals   
s8808                               Sci-fi, Horror, Action   
s8809                             Drama, Romance, Thriller   

                                               description season duration1  
show_id                                                                      
s1       As her father nears the end of his life, filmm...    NaN        90  
s2       After crossing paths at a party, a Cape Town t...      2       NaN  
s3       To protect his family from a powerful drug lor...      1       NaN  
s4       Feuds, flirtations and toilet talk go down amo...      1       NaN  
s5       In a city of coaching centers known to train I...      2       NaN  
...                                                    ...    ...       ...  
s8805    Looking to survive in a world taken over by zo...    NaN        88  
s8806    Dragged 

In [12]:
# Xoá cột duration ban đầu
netflix_df = netflix_df.drop('duration', axis=1)
# Đổi tên cột duration1 tức là cột duration mới
netflix_df = netflix_df.rename(columns={'duration1': 'duration'})

In [13]:
# Kiểm tra lại kiểu dữ liệu của từng cột
netflix_df.dtypes

type                    object
title                   object
director                object
cast                    object
country                 object
date_added      datetime64[ns]
release_year             int64
rating                  object
listed_in               object
description             object
season                  object
duration                object
dtype: object

- Sau khi tách cột duration ban đầu thành 2 cột: 1 cột season dành cho số phần của các chương trình truyền hình và 1 cột duration mới dành cho thời lượng của các bộ phim (đơn vị: Phút)
- Nhưng kiểu dữ liệu ở 2 cột vẫn chưa phù hợp với ngữ nghĩa, cần chuyển thành kiểu `float64` do có tồn tại giá trị `NaN`.
    - Giải thích cho giá trị `NaN` có tồn tại trong 2 cột dữ liệu: Do đối với các bộ phim thì cột season sẽ có giá trị `NaN` vì season chỉ dành cho các bộ phim truyền hình, và ngược lại.

In [14]:
# Chuyển cột 'duration' và 'season' thành kiểu int
netflix_df['duration'] = netflix_df['duration'].astype('float')
netflix_df['season'] = netflix_df['season'].astype('float')

In [15]:
# Kiểm tra lại kiểu dữ liệu của từng cột
netflix_df.dtypes

type                    object
title                   object
director                object
cast                    object
country                 object
date_added      datetime64[ns]
release_year             int64
rating                  object
listed_in               object
description             object
season                 float64
duration               float64
dtype: object

- Như vậy, các cột dữ liệu đã có kiểu dữ liệu phù hợp với ngữ nghĩa của chúng.

### 5. Phân bố của các cột dữ liệu

#### 5.1. Cột dữ liệu dạng số

In [16]:
# Lấy ra các cột numerical
numeric_columns = netflix_df.select_dtypes(include='number').columns[:]
numeric_columns

Index(['release_year', 'season', 'duration'], dtype='object')

Ta xem xét tỉ lệ giá trị bị thiếu trong các cột:

In [17]:
percentage_missing_dic = {}
for col in numeric_columns:
    percentage_missing = netflix_df[col].isna().mean() * 100
    percentage_missing_dic[col] = percentage_missing
    print(f'Phần trăm giá trị còn thiếu của cột \'{col}\' là {percentage_missing:.2f}%')

Phần trăm giá trị còn thiếu của cột 'release_year' là 0.00%
Phần trăm giá trị còn thiếu của cột 'season' là 69.61%
Phần trăm giá trị còn thiếu của cột 'duration' là 30.42%


Vì trong dữ liệu, phim điện ảnh chỉ có thời lượng (số phút), phim truyền hình chỉ có số mùa, nên khi tách ra 2 cột sẽ tạo giá trị null 2. Với cột season, giá trị null đến từ phim điện ảnh, nó không có số mùa. Với cột duration, giá trị null đến từ phim truyền hình, nó không có thời lượng cụ thể. Để đảm bảo không đưa ra các giả định không cần thiết khi số mùa không đóng vai trò quan trọng với phim điện ảnh và thời lượng không đóng vai trò quan trọng đối với chương trình truyền hình nên những giá trị null được giữ lại.

**Tiếp theo ta xem qua mô tả của các cột có dữ liệu dạng số xem có gì bất thường không (tối thiểu, tối đa)**

In [18]:
netflix_df.describe()

release_year       season     duration
count   8806.000000  2676.000000  6127.000000
mean    2014.179877     1.764948    99.567651
std        8.820097     1.582752    28.285800
min     1925.000000     1.000000     3.000000
25%     2013.000000     1.000000    87.000000
50%     2017.000000     1.000000    98.000000
75%     2019.000000     2.000000   114.000000
max     2024.000000    17.000000   312.000000

Ta xem xét từng cột và nhận xét:
- release_year có giá trị từ 1925 đến 2024, là thời gian năm phù hợp.
- season có giá trị từ 1 đến 17, là số mùa phù hợp cho các phim truyền hình. Số mùa trung ình là 1 cho thấy rằng max khá xa so với giá trị trung bình.
- duration có giá trị từ 3 đến 312 phút, có sự chênh lệch khá lớn.
- Các cột đều có khoảng giá trị phù hợp với ý nghĩa từng cột.

#### 5.2. Cột dữ liệu dạng phân loại

Ở phần này sẽ có 2 dạng:
- Cột có giá trị đơn: type, title, director, date_added, rating, description.
- Cột có nhiều giá trị: cast, country, listed_in.

**Xử lý cột giá trị đơn**

In [ ]:
# Lấy các cột đơn giá trí
single_columns = ['type', 'title', 'director', 'date_added', 'rating', 'description']

In [ ]:
# Tính giá trị thiếu cho các cột đơn giá trị
for col in single_columns:
    percentage_missing = netflix_df[col].isna().mean() * 100
    percentage_missing_dic[col] = percentage_missing
    print(f'Phần trăm giá trị còn thiếu của cột \'{col}\' là {percentage_missing:.2f}%')

Phần trăm giá trị còn thiếu của cột 'type' là 0.00%
Phần trăm giá trị còn thiếu của cột 'title' là 0.00%
Phần trăm giá trị còn thiếu của cột 'director' là 29.90%
Phần trăm giá trị còn thiếu của cột 'date_added' là 0.11%
Phần trăm giá trị còn thiếu của cột 'rating' là 0.05%
Phần trăm giá trị còn thiếu của cột 'description' là 0.00%


Ta thấy các cột này có giá trị thiếu rất ít có thể do các dữ liệu này là các thông tin cơ bản của một nội dung giải trí nên hầu như tất cả các nội dung được sản xuất đều phải có các thông tin này.

In [ ]:
# Lưu số lượng các giá trị khác nhau và đếm số lượng từng loại single_cate_profile
single_cate_profile = pd.DataFrame(index=["num_diff_vals", "distribution"])
num_diff_vals = netflix_df[single_columns].nunique()
distribution = netflix_df[single_columns].agg(lambda column: pd.Series(column).value_counts().to_dict())
single_cate_profile[single_columns] = [num_diff_vals, distribution]

single_cate_profile

type  \
num_diff_vals                                 2   
distribution   {'Movie': 6130, 'TV Show': 2676}   

                                                           title  \
num_diff_vals                                               8806   
distribution   {'Dick Johnson Is Dead': 1, 'Ip Man': 1, 'Care...   

                                                        director  \
num_diff_vals                                               4529   
distribution   {'Rajiv Chilaka': 19, 'Raúl Campos, Jan Suter...   

                                                      date_added  \
num_diff_vals                                               1715   
distribution   {2020-01-01 00:00:00: 110, 2019-11-01 00:00:00...   

                                                          rating  \
num_diff_vals                                                 18   
distribution   {'TV-MA': 3208, 'TV-14': 2158, 'TV-PG': 863, '...   

                                                     description  
num_diff_vals                                               8777  
distribution   {'Paranormal activity at a lush, abandoned pro...

Các cột đều có giá trị không có gì bất thường.

Có tổng cộng 2 loại hình, 8806 tiêu đề, 4529 đạo diễn, 1715 ngày thêm, 18 rating, 8777 mô tả.

**Xử lý cột đa giá trị**

In [27]:
# Lấy các cột đa giá trị
multi_columns = ['cast', 'country', 'listed_in']

In [28]:
# Tính giá trị thiếu của các cột đa giá trị
for col in multi_columns:
    percentage_missing = netflix_df[col].isna().mean() * 100
    percentage_missing_dic[col] = percentage_missing
    print(f'Phần trăm giá trị còn thiếu của cột \'{col}\' là {percentage_missing:.2f}%')

Phần trăm giá trị còn thiếu của cột 'cast' là 9.37%
Phần trăm giá trị còn thiếu của cột 'country' là 9.43%
Phần trăm giá trị còn thiếu của cột 'listed_in' là 0.00%


Ta thấy các cột này có giá trị thiếu rất ít.

In [29]:
# Lưu số lượng giá trị khác nhau và đếm số lượng từng loại và lưu vào mảng

# Tạo dataframe kết quả
def split_and_melt(column):
    df_split = column.str.split(', ', expand=True).melt().drop('variable', axis=1)
    return df_split['value']
long_df = netflix_df[multi_columns].agg(split_and_melt)
distribution = long_df.agg(lambda column: pd.Series(column).value_counts().to_dict())
num_diff_vals = long_df.nunique()

# Lưu và đếm
multiple_cate_profile = pd.DataFrame(index=["num_diff_vals", "distribution"])
multiple_cate_profile[multi_columns] = [num_diff_vals, distribution]

multiple_cate_profile

cast  \
num_diff_vals                                              36466   
distribution   {'Anupam Kher': 43, 'Shah Rukh Khan': 35, 'Jul...   

                                                         country  \
num_diff_vals                                                127   
distribution   {'United States': 3689, 'India': 1045, 'United...   

                                                       listed_in  
num_diff_vals                                                 48  
distribution   {'International Movies': 2752, 'Dramas': 2425,...

Ta có thể trích xuất cột thể loại để xem thử

In [30]:
multiple_cate_profile.loc['distribution','listed_in']

{'International Movies': 2752,
 'Dramas': 2425,
 'Comedies': 1673,
 'International TV Shows': 1350,
 'Documentaries': 869,
 'Action & Adventure': 859,
 'TV Dramas': 762,
 'Independent Movies': 755,
 'Children & Family Movies': 641,
 'Romantic Movies': 616,
 'TV Comedies': 581,
 'Thrillers': 576,
 'Crime TV Shows': 470,
 "Kids' TV": 451,
 'Docuseries': 395,
 'Music & Musicals': 375,
 'Romantic TV Shows': 370,
 'Horror Movies': 357,
 'Stand-Up Comedy': 343,
 'Reality TV': 255,
 'British TV Shows': 253,
 'Sci-Fi & Fantasy': 243,
 'Sports Movies': 219,
 'Anime Series': 176,
 'Spanish-Language TV Shows': 174,
 'TV Action & Adventure': 168,
 'Korean TV Shows': 151,
 'Classic Movies': 116,
 'LGBTQ Movies': 102,
 'TV Mysteries': 98,
 'Science & Nature TV': 92,
 'TV Sci-Fi & Fantasy': 84,
 'TV Horror': 75,
 'Cult Movies': 71,
 'Anime Features': 71,
 'Teen TV Shows': 69,
 'Faith & Spirituality': 65,
 'Movies': 57,
 'TV Thrillers': 57,
 'Stand-Up Comedy & Talk Shows': 56,
 'Classic & Cult TV': 28

Dữ liệu không có giá trị bất thường.

Có tổng cộng 36466 diễn viên, 127 quốc gia, và 48 thể loại.

#### 5.3. Nhận xét chung

- Các giá trị của các cột và các hàng sau khi tiền xử lý không có gì bất thường, các con số đều phù hợp với ý nghĩa của từng cột dữ liệu.
- Dữ liệu thiếu ở cột duration và season là do việc phân loại, những hàng có loại không quan trọng thuộc tính này thì nó sẽ để null.
- Đa số các cột có dữ liệu thiếu ít.
- Các cột cũng có các khoảng giá trị phù hợp tên cột.

# 4. Xuất file

In [ ]:
netflix_df.to_csv('netflix_titles_cleaned.csv', index=False)